examples/on_saved/src/runner.py

```diff
@@ -0,0 +1,96 @@
+import argparse
+import os
+from pathlib import Path
+
+from datasets import Dataset, load_dataset
+from torch.utils.data import DataLoader
+from training.training import tokenize_and_align_labels
+from transformers import (
+    AutoModelForTokenClassification,
+    AutoTokenizer,
+    DataCollatorForTokenClassification,
+)
+from transformers.tokenization_utils_base import PreTrainedTokenizerBase
+
+import nnbench
+from nnbench import reporter, types
+
+
+class ConllppLoader(types.ArtifactLoader):
+    def __init__(self, path: str, split: str) -> None:
+        self.path = path
+        self.split = split
+
+    def load(self) -> os.PathLike[str]:
+        dataset = load_dataset(self.path)
+        path = dataset.cache_files[self.split][0]["filename"]
+        return path
+
+
+class TokenClassificationModel(types.Artifact):
+    def deserialize(self) -> None:
+        path = Path(self.path).resolve()
+        model: AutoModelForTokenClassification = AutoModelForTokenClassification.from_pretrained(
+            path, use_safetensors=True
+        )
+        tokenizer: PreTrainedTokenizerBase = AutoTokenizer.from_pretrained(path)
+        self._value = (model, tokenizer)
+
+
+class ConllValidationData(types.Artifact):
+    def deserialize(self) -> None:
+        dataset = Dataset.from_file(self.path)
+        label_names = dataset.features["ner_tags"].feature.names
+        id2label = {i: label for i, label in enumerate(label_names)}
+        self._value = dataset, id2label
+
+
+def main(model_paths: list[str]) -> None:
+    console_reporter = reporter.BenchmarkReporter()
+    runner = nnbench.BenchmarkRunner()
+
+    models = [
+        TokenClassificationModel(loader=types.LocalArtifactLoader(path)) for path in model_paths
+    ]
+
+    model_store = types.ArtifactCollection(*models)
+
+    conllpp = ConllValidationData(ConllppLoader("conllpp", split="validation"))
+    conllpp.deserialize()
+
+    for model, tokenizer in model_store.values():
+        dataset, id2label = conllpp.value
+        tokenized_dataset = dataset.map(
+            lambda examples: tokenize_and_align_labels(tokenizer, examples),
+            batched=True,
+            remove_columns=dataset.column_names,
+        )
+        dataloader = DataLoader(
+            tokenized_dataset,
+            shuffle=False,
+            collate_fn=DataCollatorForTokenClassification(tokenizer, padding=True),
+            batch_size=8,
+        )
+        result = runner.run(
+            "benchmark.py",
```

Comment:
I would be interested in what happens here - does the runner collect and clear everytime or does the collection happen once and is then reused in the process?

Reply:
Collection happens only in the first run :)